File: macos/Sources/ElevateCore/Models/ActivationProfile.swift
import Foundation

/// Where a profile came from. User profiles live in `state.json`; managed ones are published by
/// the organization and resolved at runtime, so they are never persisted.
public enum ProfileSource: String, Codable, Hashable, Sendable { case user, managed }

/// A named set of roles and groups activated together, across accounts and tenants.
public struct ActivationProfile: Codable, Hashable, Sendable, Identifiable {
    public struct Entry: Codable, Hashable, Sendable {
        public var roleKey: RoleKey
        /// Duration used on the last run of this entry; nil until the profile has run.
        public var lastDuration: Duration?
        public init(roleKey: RoleKey, lastDuration: Duration? = nil) {
            self.roleKey = roleKey
            self.lastDuration = lastDuration
        }
    }

    public var id: UUID
    public var name: String
    public var entries: [Entry]
    /// Reason entered on the last run; prefilled next time.
    public var lastJustification: String?
    /// Shown as a chip in the panel. At most `ProfilePins.limit` profiles are pinned at a time.
    public var pinned: Bool
    /// `.managed` profiles are published by the organization: read-only in the UI and never saved.
    public var source: ProfileSource

    public init(id: UUID = UUID(), name: String, entries: [Entry], lastJustification: String? = nil,
                pinned: Bool = false, source: ProfileSource = .user) {
        self.id = id
        self.name = name
        self.entries = entries
        self.lastJustification = lastJustification
        self.pinned = pinned
        self.source = source
    }

    private enum CodingKeys: String, CodingKey {
        case id, name, entries, lastJustification, pinned, source
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(UUID.self, forKey: .id)
        name = try c.decode(String.self, forKey: .name)
        entries = try c.decode([Entry].self, forKey: .entries)
        lastJustification = try c.decodeIfPresent(String.self, forKey: .lastJustification)
        pinned = try c.decodeIfPresent(Bool.self, forKey: .pinned) ?? false
        source = try c.decodeIfPresent(ProfileSource.self, forKey: .source) ?? .user
    }

    /// `pinned` and `source` are written only when they are not the default, so files from before
    /// those fields existed round-trip unchanged.
    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(name, forKey: .name)
        try c.encode(entries, forKey: .entries)
        try c.encodeIfPresent(lastJustification, forKey: .lastJustification)
        if pinned { try c.encode(true, forKey: .pinned) }
        if source != .user { try c.encode(source, forKey: .source) }
    }
}

public enum ProfilePins {
    /// How many profiles may be pinned to the panel; one row of chips that never wraps.
    public static let limit = 4
}

public enum ProfileSummary {
    /// "3 roles · 1 group" style caption for a chip. Entra and Azure count as roles.
    public static func caption(entries: [ActivationProfile.Entry]) -> String {
        let groups = entries.filter { $0.roleKey.scope.kind == .group }.count
        let roles = entries.count - groups
        var parts: [String] = []
        if roles > 0 { parts.append("\(roles) role\(roles == 1 ? "" : "s")") }
        if groups > 0 { parts.append("\(groups) group\(groups == 1 ? "" : "s")") }
        return parts.isEmpty ? "empty" : parts.joined(separator: " · ")
    }
}

File: windows/src/Elevate.Core/Models/ActivationProfile.cs
using System.Text.Json.Serialization;

namespace Elevate.Core.Models;

/// <summary>
/// Where a profile came from. User profiles live in <c>state.json</c>; managed ones are published
/// by the organization and resolved at runtime, so they are never persisted.
/// </summary>
public enum ProfileSource { User, Managed }

/// <summary>A named set of roles and groups activated together, across accounts and tenants.</summary>
public sealed record ActivationProfile
{
    [JsonConstructor]
    public ActivationProfile(Guid id, string name, List<Entry> entries, string? lastJustification = null, bool pinned = false)
    {
        ArgumentNullException.ThrowIfNull(entries);
        Id = id;
        Name = name;
        Entries = entries;
        LastJustification = lastJustification;
        Pinned = pinned;
    }

    /// <summary>A new profile with a fresh id.</summary>
    public ActivationProfile(string name, IEnumerable<Entry> entries, string? lastJustification = null)
        : this(Guid.NewGuid(), name, [.. entries ?? []], lastJustification)
    {
    }

    /// <summary>One role in a profile; <see cref="Entry.LastDuration"/> is null until the profile has run.</summary>
    public sealed record Entry(RoleKey RoleKey, TimeSpan? LastDuration = null);

    public Guid Id { get; set; }

    public string Name { get; set; }

    public List<Entry> Entries { get; set; }

    /// <summary>Reason entered on the last run; prefilled next time.</summary>
    public string? LastJustification { get; set; }

    /// <summary>
    /// Shown as a chip in the panel; at most <see cref="ProfilePins.Limit"/> profiles are pinned at a
    /// time. Written only when true, so files from before pinning existed round-trip unchanged.
    /// </summary>
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public bool Pinned { get; set; }

    /// <summary>
    /// <see cref="ProfileSource.Managed"/> profiles are published by the organization: read-only in
    /// the UI and never saved. Written only when managed, so a user profile's JSON is unchanged.
    /// </summary>
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public ProfileSource Source { get; init; } = ProfileSource.User;

    /// <summary>
    /// A copy with its own <see cref="Entries"/> list, so mutating one profile's entries does not
    /// touch the other's. <see cref="Entry"/> is an immutable record, so the entries are shared.
    /// (Records may not declare a member named <c>Clone</c>, hence the name.)
    /// </summary>
    public ActivationProfile DeepCopy() => new(Id, Name, [.. Entries], LastJustification, Pinned) { Source = Source };

    public bool Equals(ActivationProfile? other) =>
        other is not null
        && Id == other.Id
        && Name == other.Name
        && LastJustification == other.LastJustification
        && Pinned == other.Pinned
        && Source == other.Source
        && Entries.SequenceEqual(other.Entries);

    public override int GetHashCode() => HashCode.Combine(Id, Name, Entries.Count, LastJustification, Pinned, Source);
}

public static class ProfilePins
{
    /// <summary>How many profiles may be pinned to the panel; one row of chips that never wraps.</summary>
    public const int Limit = 4;
}

public static class ProfileSummary
{
    /// <summary>"3 roles · 1 group" style caption for a chip. Entra and Azure count as roles.</summary>
    public static string Caption(IEnumerable<ActivationProfile.Entry> entries)
    {
        ArgumentNullException.ThrowIfNull(entries);

        var all = entries.ToList();
        var groups = all.Count(e => e.RoleKey.Scope.Kind == RoleScopeKind.Group);
        var roles = all.Count - groups;
        var parts = new List<string>();
        if (roles > 0)
        {
            parts.Add($"{roles} role{(roles == 1 ? "" : "s")}");
        }

        if (groups > 0)
        {
            parts.Add($"{groups} group{(groups == 1 ? "" : "s")}");
        }

        return parts.Count == 0 ? "empty" : string.Join(" · ", parts);
    }
}
